File: connection/cluster/ClusterSession.ts
/*
 * Copyright (C) 2021 Vaticle
 *
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import { Database } from "../../api/connection/database/Database";
import { TypeDBClusterOptions, TypeDBOptions } from "../../api/connection/TypeDBOptions";
import { SessionType, TypeDBSession } from "../../api/connection/TypeDBSession";
import { TransactionType, TypeDBTransaction } from "../../api/connection/TypeDBTransaction";
import { TypeDBSessionImpl } from "../TypeDBSessionImpl";
import { ClusterClient } from "./ClusterClient";
import { ClusterServerClient } from "./ClusterServerClient";
import { FailsafeTask } from "./FailsafeTask";

export class ClusterSession implements TypeDBSession {

    private readonly _clusterClient: ClusterClient;
    private _options: TypeDBClusterOptions;

    constructor(clusterClient: ClusterClient, serverAddress: string) {
        this._clusterClient = clusterClient;
        this._clusterServerClient = clusterClient.clusterServerClient(serverAddress.toString());
    }

    private _clusterServerClient: ClusterServerClient;

    get clusterServerClient(): ClusterServerClient {
        return this._clusterServerClient;
    }

    set clusterServerClient(client: ClusterServerClient) {
        this._clusterServerClient = client;
    }

    private _typeDBSession: TypeDBSessionImpl;

    get typeDBSession(): TypeDBSessionImpl {
        return this._typeDBSession;
    }

    set typeDBSession(session: TypeDBSessionImpl) {
        this._typeDBSession = session;
    }

    async open(serverAddress: string, database: string, type: SessionType, options: TypeDBClusterOptions): Promise<ClusterSession> {
        console.info(`Opening a session to '${serverAddress}'`);
        this._typeDBSession = await this._clusterServerClient.session(database, type, options);
        this._options = options;
        return this;
    }

    transaction(type: TransactionType, options: TypeDBClusterOptions = TypeDBOptions.cluster()): Promise<TypeDBTransaction> {
        if (options.readAnyReplica) {
            return this.transactionAnyReplica(type, options);
        } else {
            return this.transactionPrimaryReplica(type, options);
        }
    }

    type(): SessionType {
        return this._typeDBSession.type();
    }

    isOpen(): boolean {
        return this._typeDBSession.isOpen();
    }

    options(): TypeDBClusterOptions {
        return this._options;
    }

    close(): Promise<void> {
        return this._typeDBSession.close();
    }

    database(): Database {
        return this._typeDBSession.database();
    }

    clusterClient(): ClusterClient {
        return this._clusterClient;
    }

    private transactionPrimaryReplica(type: TransactionType, options: TypeDBClusterOptions): Promise<TypeDBTransaction> {
        return new TransactionFailsafeTask(this, type, options).runPrimaryReplica();
    }

    private transactionAnyReplica(type: TransactionType, options: TypeDBClusterOptions): Promise<TypeDBTransaction> {
        return new TransactionFailsafeTask(this, type, options).runAnyReplica();
    }
}

class TransactionFailsafeTask extends FailsafeTask<TypeDBTransaction> {
    private readonly _clusterSession: ClusterSession;
    private readonly _type: TransactionType;
    private readonly _options: TypeDBClusterOptions;

    constructor(clusterSession: ClusterSession, type: TransactionType, options: TypeDBClusterOptions) {
        super(clusterSession.clusterClient(), clusterSession.database().name());
        this._clusterSession = clusterSession;
        this._type = type;
        this._options = options;
    }

    run(replica: Database.Replica): Promise<TypeDBTransaction> {
        return this._clusterSession.typeDBSession.transaction(this._type, this._options);
    }

    async rerun(replica: Database.Replica): Promise<TypeDBTransaction> {
        if (this._clusterSession.typeDBSession) await this._clusterSession.typeDBSession.close();
        this._clusterSession.clusterServerClient = this._clusterSession.clusterClient().clusterServerClient(replica.address());
        this._clusterSession.typeDBSession = await this._clusterSession.clusterServerClient.session(this.database, this._clusterSession.type(), this._clusterSession.options());
        return await this._clusterSession.typeDBSession.transaction(this._type, this._options);
    }
}

File: connection/cluster/ClusterSession.java
/*
 * Copyright (C) 2022 Vaticle
 *
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package com.vaticle.typedb.client.connection.cluster;

import com.vaticle.typedb.client.api.TypeDBOptions;
import com.vaticle.typedb.client.api.TypeDBSession;
import com.vaticle.typedb.client.api.TypeDBTransaction;
import com.vaticle.typedb.client.api.database.Database;
import com.vaticle.typedb.client.connection.TypeDBSessionImpl;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class ClusterSession implements TypeDBSession {

    private static final Logger LOG = LoggerFactory.getLogger(TypeDBSession.class);
    private final ClusterClient clusterClient;
    private final TypeDBOptions.Cluster options;
    private ClusterServerClient clusterServerClient;
    private TypeDBSessionImpl typeDBSession;

    public ClusterSession(ClusterClient clusterClient, String serverAddress, String database, Type type, TypeDBOptions.Cluster options) {
        this.clusterClient = clusterClient;
        this.clusterServerClient = clusterClient.clusterServerClient(serverAddress);
        LOG.debug("Opening a session to '{}'", serverAddress);
        this.typeDBSession = clusterServerClient.session(database, type, options);
        this.options = options;
    }

    @Override
    public TypeDBTransaction transaction(TypeDBTransaction.Type type) {
        return transaction(type, TypeDBOptions.cluster());
    }

    @Override
    public TypeDBTransaction transaction(TypeDBTransaction.Type type, TypeDBOptions options) {
        TypeDBOptions.Cluster clusterOpt = options.asCluster();
        if (clusterOpt.readAnyReplica().isPresent() && clusterOpt.readAnyReplica().get()) {
            return transactionAnyReplica(type, clusterOpt);
        } else {
            return transactionPrimaryReplica(type, options);
        }
    }

    private TypeDBTransaction transactionPrimaryReplica(TypeDBTransaction.Type type, TypeDBOptions options) {
        return transactionFailsafeTask(type, options).runPrimaryReplica();
    }

    private TypeDBTransaction transactionAnyReplica(TypeDBTransaction.Type type, TypeDBOptions.Cluster options) {
        return transactionFailsafeTask(type, options).runAnyReplica();
    }

    private ClusterClient.FailsafeTask<TypeDBTransaction> transactionFailsafeTask(TypeDBTransaction.Type type, TypeDBOptions options) {
        return clusterClient.createFailsafeTask(
                database().name(),
                (parameter) -> typeDBSession.transaction(type, options),
                (parameter) -> {
                    if (typeDBSession != null) typeDBSession.close();
                    clusterServerClient = parameter.client();
                    typeDBSession = clusterServerClient.session(database().name(), ClusterSession.this.type(), ClusterSession.this.options());
                    return typeDBSession.transaction(type, options);
                }
        );
    }

    @Override
    public TypeDBSession.Type type() {
        return typeDBSession.type();
    }

    @Override
    public TypeDBOptions.Cluster options() {
        return options;
    }

    @Override
    public boolean isOpen() {
        return typeDBSession.isOpen();
    }

    @Override
    public void onClose(Runnable function) {
        typeDBSession.onClose(function);
    }

    @Override
    public void close() {
        typeDBSession.close();
    }

    @Override
    public Database database() {
        return typeDBSession.database();
    }
}
